hogql_parser/parser.cpp

```diff
@@ -0,0 +1,1358 @@
+    }
+  }
+
+  PyObject* visitAsPyObject(antlr4::tree::ParseTree* tree) {
+    PyObject* cast_result = any_cast<PyObject*>(visit(tree));
+    if (!cast_result) {
+      throw runtime_error("Rule resulted in a null PyObject pointer. A Python exception must be set at this point.");
+    }
+    return cast_result;
+  }
+
+  PyObject* visitAsPyObjectOrNone(antlr4::tree::ParseTree* tree) {
+    if (tree == NULL) {
+      Py_RETURN_NONE;
+    }
+    return visitAsPyObject(tree);
+  }
+
+  PyObject* visitAsPyObjectOrEmptyList(antlr4::tree::ParseTree* tree) {
+    if (tree == NULL) {
+      return PyList_New(0);
+    }
+    return visitAsPyObject(tree);
+  }
+
+  // T has to be used in place of antlr4::tree::ParseTree* here, because there's no conversion from the child class
+  // to its parent within vectors
+  template <typename T>
+  PyObject* visitPyListOfObjects(vector<T> tree) {
+    PyObject* result = PyList_New(tree.size());
+    for (size_t i = 0; i < tree.size(); i++) {
+      PyList_SET_ITEM(result, i, visitAsPyObject(tree[i]));
+    }
+    return result;
+  }
+
+  string visitAsString(antlr4::tree::ParseTree* tree) { return any_cast<string>(visit(tree)); }
+
+  template <typename T>
+  vector<string> visitAsVectorOfStrings(vector<T> tree) {
+    vector<string> result;
+    result.reserve(tree.size());
+    for (auto child : tree) {
+      result.push_back(visitAsString(child));
+    }
+    return result;
+  }
+
+  VISIT(Select) {
+    auto select_union_stmt_ctx = ctx->selectUnionStmt();
+    if (select_union_stmt_ctx) {
+      return visit(select_union_stmt_ctx);
+    }
+    return visit(ctx->selectStmt());
+  }
+
+  VISIT(SelectStmtWithParens) {
+    auto select_stmt_ctx = ctx->selectStmt();
+    if (select_stmt_ctx) {
+      return visit(select_stmt_ctx);
+    }
+    return visit(ctx->selectUnionStmt());
+  }
+
+  VISIT(SelectUnionStmt) {
+    // Using a vector of PyObjects atypically here, because this is a precursor of flattened_queries
+    vector<PyObject*> select_queries;
+    auto select_stmt_with_parens_ctxs = ctx->selectStmtWithParens();
+    select_queries.reserve(select_stmt_with_parens_ctxs.size());
+    for (auto select_stmt_with_parens_ctx : select_stmt_with_parens_ctxs) {
+      select_queries.push_back(visitAsPyObject(select_stmt_with_parens_ctx));
+    }
+    PyObject* flattened_queries = PyList_New(0);
+    for (auto query : select_queries) {
+      if (is_ast_node_instance(query, "SelectQuery")) {
+        PyList_Append(flattened_queries, query);
+      } else if (is_ast_node_instance(query, "SelectUnionQuery")) {
+        // Extend flattened_queries with sub_select_queries
+        PyObject* sub_select_queries = PyObject_GetAttrString(query, "select_queries");
+        X_PyList_Extend(flattened_queries, sub_select_queries);
+        Py_DECREF(sub_select_queries);
+      } else {
+        Py_DECREF(flattened_queries);
+        throw HogQLParsingException("Unexpected query node type: " + string(Py_TYPE(query)->tp_name));
```

Comment:
Should anything be `DECREF`-d on the `select_queries` when this throws?

Reply:
Good catch, yeah, they all should be decremented

Reply:
I'll do the same thing as with the other missing DECREFs noted above – follow-up PR tightening that up